### PR TITLE
fix(apiserver,volume) Removing PVC dependency from volume creation path

### PR DESCRIPTION
**Pull request by mynktl**
Changes:
- This PR **remove PVC dependency** **(used earlier for CStor replica affinity #966, #950)** from Volume creation path in maya-apiserver **if PV creation request has annotation `openebs.io/created-through=restore` set.**

 
Signed-off-by: mayank <mayank.patel@mayadata.io>

### CODE REVIEW

**Comment by kmova** [comment]
As discussed in the Jan 8 2020 standup, moved this to 1.7.0 release as it needs further beta testing.

File: pkg/install/v1alpha1/cstor_volume.go [hunk 1]
```diff
@@ -110,7 +110,6 @@ spec:
   run:
     tasks:
     - cstor-volume-create-getstorageclass-default
-    - cstor-volume-create-getpvc-default
```
Comment (vishnuitta) [inline pkg/install/v1alpha1/cstor_volume.go]: not calling this is going to impact current functionality which depends on PVC..
let us still call this and set the values to default as 'none' (or empty) by handling the error from 'get' operation

File: pkg/volume/volume.go [hunk 2]
```diff
@@ -113,20 +109,6 @@ func (v *Operation) Create() (*v1alpha1.CASVolume, error) {
 		return nil, errors.Errorf("failed to create volume: missing volume capacity: %s", v.volume)
 	}
 
-	pvcName := v.volume.Labels[string(v1alpha1.PersistentVolumeClaimKey)]
-	if len(pvcName) == 0 {
```
Comment (vishnuitta) [inline pkg/volume/volume.go]: instead of removing, lets handle the case when the request doesn't have this label

**Comment by AmitKumarDas** [comment]
@mynktl Can we put more details on what was the previous usage & why it has been removed now. How will the feature do without this feature & so on.

File: pkg/install/v1alpha1/cstor_volume.go [hunk 3]
```diff
@@ -699,8 +695,8 @@ spec:
       {{- end }}
       {{- end }}
   task: |
-    {{- $replicaAntiAffinity := .TaskResult.creategetpvc.replicaAntiAffinity -}}
-    {{- $preferredReplicaAntiAffinity := .TaskResult.creategetpvc.preferredReplicaAntiAffinity }}
+    {{- $replicaAntiAffinity := .TaskResult.creategetpvc.replicaAntiAffinity | default "" -}}
```
Comment (AmitKumarDas) [inline pkg/install/v1alpha1/cstor_volume.go]: If these are not used; then why not remove them altogether?

**Comment by mynktl** [comment]
@AmitKumarDas AFAIK we are using PVC details here for replica affinity. cc: @vishnuitta @prateekpandey14 

We are removing this PVC dependency here to support clone in different namespace through velero -- this operation doesn't have PVC details. It will create PV first then it creates PVC with PV details.

**Comment by mynktl** [comment]
@AmitKumarDas    #966  has added above feature.

File: pkg/install/v1alpha1/cstor_volume.go (hunk 1 above)
Comment (mynktl) [reply pkg/install/v1alpha1/cstor_volume.go]: done

File: pkg/volume/volume.go (hunk 2 above)
Comment (mynktl) [reply pkg/volume/volume.go]: done

**Comment by mynktl** [comment]
@kmova 
I am seeing that most of the cases that you have mentioned are already covered(https://github.com/openebs/maya/blob/master/tests/cstor/snapshot/snapshot_test.go  https://github.com/openebs/maya/blob/master/tests/cstor/clone/clone_test.go)
Only missing case is deleting parent PVC prior to clone PVC. I will update the BDD for this.

**Comment by mynktl** [comment]
logs for BDD to create CStorVolume without PVC
```
Running Suite: Test cstor volume provisioning
=============================================
Random Seed: 1583383344
Will run 3 of 13 specs

STEP: waiting for maya-apiserver pod to come into running state
STEP: waiting for openebs-provisioner pod to come into running state
STEP: Verifying 'admission-server' pod status as running
STEP: building a namespace
STEP: creating a namespace
SSSSS
------------------------------
[Cstor Volume Provisioning Positive] TEST VOLUME PROVISIONING when cstor pvc with replicacount n is created 
  should create cstor volume target pod
  /home/mayank/go/src/github.com/openebs/maya/tests/cstor/volume/provision_test.go:106
STEP: building spc object
STEP: creating storagepoolclaim
STEP: verifying healthy csp count
STEP: building a CAS Config with generated SPC name
STEP: building object of storageclass
STEP: creating storageclass
STEP: building a pvc
STEP: creating above pvc
STEP: verifying target pod count as 1
STEP: verifying cstorvolume replica count
STEP: verifying pvc status as bound
STEP: should delete cstor pools pod before pvc delete
STEP: deleting above pvc
STEP: verifying target pod count as 0
STEP: verifying deleted pvc
STEP: verifying if cstorvolume is deleted
STEP: deleting resources created for testing cstor volume provisioning
STEP: deleting storagepoolclaim
STEP: deleting storageclass

• [SLOW TEST:121.195 seconds]
[Cstor Volume Provisioning Positive] TEST VOLUME PROVISIONING
/home/mayank/go/src/github.com/openebs/maya/tests/cstor/volume/provision_test.go:47
  when cstor pvc with replicacount n is created
  /home/mayank/go/src/github.com/openebs/maya/tests/cstor/volume/provision_test.go:105
    should create cstor volume target pod
    /home/mayank/go/src/github.com/openebs/maya/tests/cstor/volume/provision_test.go:106
------------------------------
[Cstor Volume Provisioning Positive] TEST VOLUME PROVISIONING when cstor PV with replicacount n is created without PVC 
  should create cstor volume target pod
  /home/mayank/go/src/github.com/openebs/maya/tests/cstor/volume/provision_test.go:206
STEP: building spc object
STEP: creating storagepoolclaim
STEP: verifying healthy csp count
STEP: building a CAS Config with generated SPC name
STEP: building object of storageclass
STEP: creating storageclass
STEP: verifying target pod count as 1
STEP: verifying cstorvolume replica count
STEP: verifying target pod count as 0
STEP: verifying cstorvolume replica count
STEP: deleting resources created for testing cstor volume provisioning
STEP: deleting storagepoolclaim
STEP: deleting storageclass

• [SLOW TEST:79.053 seconds]
[Cstor Volume Provisioning Positive] TEST VOLUME PROVISIONING
/home/mayank/go/src/github.com/openebs/maya/tests/cstor/volume/provision_test.go:47
  when cstor PV with replicacount n is created without PVC
  /home/mayank/go/src/github.com/openebs/maya/tests/cstor/volume/provision_test.go:205
    should create cstor volume target pod
    /home/mayank/go/src/github.com/openebs/maya/tests/cstor/volume/provision_test.go:206
------------------------------
[Cstor Volume Provisioning Positive] TEST VOLUME PROVISIONING when cstor PV with replicacount n is created without PVC and restore label 
  should create cstor volume target pod
  /home/mayank/go/src/github.com/openebs/maya/tests/cstor/volume/provision_test.go:278
STEP: building spc object
STEP: creating storagepoolclaim
STEP: verifying healthy csp count
STEP: building a CAS Config with generated SPC name
STEP: building object of storageclass
STEP: creating storageclass
STEP: verifying target pod count as 1
STEP: verifying cstorvolume replica count
STEP: verifying target pod count as 0
STEP: verifying cstorvolume replica count
STEP: deleting resources created for testing cstor volume provisioning
STEP: deleting storagepoolclaim
STEP: deleting storageclass

• [SLOW TEST:53.487 seconds]
[Cstor Volume Provisioning Positive] TEST VOLUME PROVISIONING
/home/mayank/go/src/github.com/openebs/maya/tests/cstor/volume/provision_test.go:47
  when cstor PV with replicacount n is created without PVC and restore label
  /home/mayank/go/src/github.com/openebs/maya/tests/cstor/volume/provision_test.go:277
    should create cstor volume target pod
    /home/mayank/go/src/github.com/openebs/maya/tests/cstor/volume/provision_test.go:278
------------------------------
SSSSSSTEP: deleting namespace

Ran 3 of 13 Specs in 256.630 seconds
SUCCESS! -- 3 Passed | 0 Failed | 0 Pending | 10 Skipped
PASS

Ginkgo ran 1 suite in 4m20.172353494s
Test Suite Passed
```

**Comment by mynktl** [comment]
logs for `[cstor] TEST VOLUME CLONE PROVISIONING` with deleting clone-pvc before source pvc
```
Running Suite: Test cstor volume clone provisioning
===================================================
Random Seed: 1583389225
Will run 1 of 1 specs

STEP: waiting for maya-apiserver pod to come into running state
STEP: waiting for openebs-provisioner pod to come into running state
STEP: Verifying 'admission-server' pod status as running
STEP: building a namespace
STEP: creating a namespace
[cstor] TEST VOLUME CLONE PROVISIONING when cstor pvc with replicacount 1 is created 
  should create cstor volume target pod
  /home/mayank/go/src/github.com/openebs/maya/tests/cstor/clone/clone_test.go:103
STEP: building spc object
STEP: creating storagepoolclaim
STEP: verifying healthy csp count
STEP: building a CAS Config with generated SPC name
STEP: building storageclass object
STEP: creating storageclass
STEP: building a persistentvolumeclaim
STEP: creating cstor persistentvolumeclaim
STEP: verifying volume target pod count as 1
STEP: verifying cstorvolume replica count
STEP: verifying pvc status as bound
STEP: verifying cstorVolume status as healthy
STEP: building a busybox app pod using above cstor volume
STEP: creating pod with above pvc as volume
STEP: verifying app pod is running
STEP: building a cstor volume snapshot
STEP: creating cstor volume snapshot
STEP: builing clone persistentvolumeclaim
STEP: creating clone persistentvolumeclaim
STEP: verifying clone volume target pod count
STEP: verifying clone volumeereplica count
STEP: verifying clone pvc status as bound
STEP: building a busybox app pod with above clone pvc as volume
STEP: creating pod with above clone pvc as volume
STEP: verifying pod is running
STEP: verifying data consistency in pvc and clone pvc
STEP: fetching data from original pvc
STEP: fetching data from clone pvc
STEP: veryfing data consistency
STEP: try deleting source persistentvolumeclaim
STEP: deleting application pod
STEP: deleting clone application pod
STEP: deleting clone persistentvolumeclaim
STEP: verifying clone target pod count as 0
STEP: verifying deleted clone pvc
STEP: verifying if clone cstorvolume is deleted
STEP: deleting cstor volume snapshot
STEP: verifying deleted snapshot
STEP: deleting source persistentvolumeclaim
STEP: verifying source volume target pod count as 0
STEP: verifying deleted source pvc
STEP: verifying if source cstorvolume is deleted
STEP: deleting resources created for cstor volume snapshot provisioning
STEP: deleting storagepoolclaim
STEP: deleting storageclass

• [SLOW TEST:261.226 seconds]
[cstor] TEST VOLUME CLONE PROVISIONING
/home/mayank/go/src/github.com/openebs/maya/tests/cstor/clone/clone_test.go:41
  when cstor pvc with replicacount 1 is created
  /home/mayank/go/src/github.com/openebs/maya/tests/cstor/clone/clone_test.go:102
    should create cstor volume target pod
    /home/mayank/go/src/github.com/openebs/maya/tests/cstor/clone/clone_test.go:103
------------------------------
STEP: deleting namespace

Ran 1 of 1 Specs in 263.974 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS

Ginkgo ran 1 suite in 4m31.396516533s
Test Suite Passed
```

**Pull request merged**